File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/data/repository/CashierRepositoryImpl.java
package com.bits.bee.bpmc.data.repository;

import java.lang.System;

/**
 * Created by aldi on 17/03/22.
 */
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000D\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\b\n\u0000\n\u0002\u0010 \n\u0002\b\u0002\n\u0002\u0010\u0002\n\u0002\b\u0003\u0018\u00002\u00020\u0001B\u001f\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u0012\u0006\u0010\u0006\u001a\u00020\u0007\u00a2\u0006\u0002\u0010\bJ\u0010\u0010\t\u001a\n\u0012\u0006\u0012\u0004\u0018\u00010\u000b0\nH\u0016J\u001c\u0010\f\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u000b0\r0\n2\u0006\u0010\u000e\u001a\u00020\u000fH\u0016J\"\u0010\u0010\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u000b0\u00110\r0\n2\u0006\u0010\u0012\u001a\u00020\u000fH\u0016J\u0019\u0010\u0013\u001a\u00020\u00142\u0006\u0010\u0015\u001a\u00020\u000bH\u0096@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\u0016R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0006\u001a\u00020\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000\u0082\u0002\u0004\n\u0002\b\u0019\u00a8\u0006\u0017"}, d2 = {"Lcom/bits/bee/bpmc/data/repository/CashierRepositoryImpl;", "Lcom/bits/bee/bpmc/domain/repository/CashierRepository;", "apiUtils", "Lcom/bits/bee/bpmc/data/data_source/remote/ApiUtils;", "cashierDao", "Lcom/bits/bee/bpmc/data/data_source/local/dao/CashierDao;", "defaultDispatcher", "Lkotlinx/coroutines/CoroutineDispatcher;", "(Lcom/bits/bee/bpmc/data/data_source/remote/ApiUtils;Lcom/bits/bee/bpmc/data/data_source/local/dao/CashierDao;Lkotlinx/coroutines/CoroutineDispatcher;)V", "getActiveCashier", "Lkotlinx/coroutines/flow/Flow;", "Lcom/bits/bee/bpmc/domain/model/Cashier;", "getCashierById", "Lcom/bits/bee/bpmc/utils/Resource;", "id", "", "getCashierList", "", "branchId", "updateActiveCashier", "", "cashier", "(Lcom/bits/bee/bpmc/domain/model/Cashier;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "app_debug"})
public final class CashierRepositoryImpl implements com.bits.bee.bpmc.domain.repository.CashierRepository {
    private final com.bits.bee.bpmc.data.data_source.remote.ApiUtils apiUtils = null;
    private final com.bits.bee.bpmc.data.data_source.local.dao.CashierDao cashierDao = null;
    private final kotlinx.coroutines.CoroutineDispatcher defaultDispatcher = null;
    
    @javax.inject.Inject()
    public CashierRepositoryImpl(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.remote.ApiUtils apiUtils, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.local.dao.CashierDao cashierDao, @org.jetbrains.annotations.NotNull()
    kotlinx.coroutines.CoroutineDispatcher defaultDispatcher) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.Cashier>>> getCashierList(int branchId) {
        return null;
    }
    
    @org.jetbrains.annotations.Nullable()
    @java.lang.Override()
    public java.lang.Object updateActiveCashier(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.Cashier cashier, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.domain.model.Cashier> getActiveCashier() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<com.bits.bee.bpmc.domain.model.Cashier>> getCashierById(int id) {
        return null;
    }
}

File: app/src/main/java/com/bits/bee/bpmc/data/repository/CashierRepositoryImpl.kt
package com.bits.bee.bpmc.data.repository

import com.bits.bee.bpmc.data.data_source.local.dao.CashierDao
import com.bits.bee.bpmc.data.data_source.remote.ApiUtils
import com.bits.bee.bpmc.data.data_source.remote.model.CashierPost
import com.bits.bee.bpmc.data.data_source.remote.model.MonitCashierPost
import com.bits.bee.bpmc.data.data_source.remote.response.CashierResponse
import com.bits.bee.bpmc.data.data_source.remote.response.CashierStatusResponse
import com.bits.bee.bpmc.domain.mapper.CashierDataMapper
import com.bits.bee.bpmc.domain.model.Cashier
import com.bits.bee.bpmc.domain.repository.CashierRepository
import com.bits.bee.bpmc.utils.ApiResponse
import com.bits.bee.bpmc.utils.NetworkBoundResource
import com.bits.bee.bpmc.utils.NetworkDatabaseBoundResource
import com.bits.bee.bpmc.utils.Resource
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import okhttp3.ResponseBody
import javax.inject.Inject

/**
 * Created by aldi on 17/03/22.
 */
class CashierRepositoryImpl @Inject constructor(
    private val apiUtils: ApiUtils,
    private val cashierDao: CashierDao,
    private val defaultDispatcher: CoroutineDispatcher
) : CashierRepository {

    override fun getCashierList(branchId : Int): Flow<Resource<List<Cashier>>> {

        return object : NetworkDatabaseBoundResource<List<Cashier>, CashierResponse>(){
            override suspend fun loadFormDB(): List<Cashier> {
                return cashierDao.getListCashier().map { CashierDataMapper.fromDbToDomain(it) }
            }

            override fun shouldFetch(data: List<Cashier>?): Boolean {
                return true
            }

            override suspend fun createCall(): Flow<ApiResponse<CashierResponse>> {
                return apiUtils.getCashierApiService().getCashierList(branchId)
            }

            override suspend fun saveCallResult(data: CashierResponse) {
                cashierDao.deleteAll()
                cashierDao.insertBulk(data.data.map { CashierDataMapper.fromNetworkToDb(it) })
            }

        }.getAsFlow()
    }

    override suspend fun updateActiveCashier(cashier: Cashier) {
        withContext(defaultDispatcher){
            cashierDao.resetActive()
            cashierDao.update(CashierDataMapper.fromDomainToDb(cashier))
        }
    }

    override fun postActivateCashier(cashierPost: CashierPost): Flow<Resource<CashierStatusResponse>> {
        return object : NetworkBoundResource<CashierStatusResponse>(){
            override fun createCall(): Flow<ApiResponse<CashierStatusResponse>> {
                return apiUtils.getCashierApiService().postActivateCashier(cashierPost)
            }

        }.getAsFlow()
    }

    override fun detachActivateCashier(cashierPost: CashierPost): Flow<Resource<CashierStatusResponse>> {
        return object : NetworkBoundResource<CashierStatusResponse>(){
            override fun createCall(): Flow<ApiResponse<CashierStatusResponse>> {
                return apiUtils.getCashierApiService().postDetachCashier(cashierPost)
            }

        }.getAsFlow()
    }

    override fun getActiveCashier(): Flow<Cashier?> = flow{
        var cashier : Cashier? = null
        cashierDao.getActiveCashier()?.let {
            cashier = CashierDataMapper.fromDbToDomain(it)
        }
        emit(cashier)
    }.flowOn(defaultDispatcher)

    override fun getCashierById(id: Int): Flow<Cashier?> {
        return flow {
            val data = cashierDao.getCashierById(id)
            emit(data?.let {  CashierDataMapper.fromDbToDomain(data)})
        }.flowOn(defaultDispatcher)
    }

    override fun detachCashier(cashierPost: CashierPost): Flow<Resource<CashierStatusResponse>> {
        return object : NetworkBoundResource<CashierStatusResponse>() {
            override fun createCall(): Flow<ApiResponse<CashierStatusResponse>> {
                return apiUtils.getCashierApiService().postDetachCashier(cashierPost)
            }
        }.getAsFlow()
    }

    override fun postMonitCashier(monitCashierPost: MonitCashierPost): Flow<Resource<ResponseBody>> {
        return object : NetworkBoundResource<ResponseBody>() {
            override fun createCall(): Flow<ApiResponse<ResponseBody>> {
                return apiUtils.getCashierApiService().postMonitCashier(monitCashierPost)
            }
        }.getAsFlow()
    }

}